woocommerce/woocommerce-ios
# Shipping Labels: Custom package form validation (part 1)

Part of: #4743

## Description

This PR introduces validation for the custom package creation form, when creating a new custom package in the Shipping Label flow. Some details about the validation:

* All fields are required, and each should display a message about that if you enter invalid input / erase the input
* The name field must have content (besides whitespace/newlines)
* The dimensions fields (length/width/height) must have values greater than 0
* The weight field must have a value 0 or greater
* Once all fields have valid content, the Done button is enabled

I stopped at this point because I didn't want this PR to get too big, and I wanted to get feedback on the approach I used here. What isn't done yet:

* Right now, the only validation error message is "This field is required" (to match the web and Android clients). However, this doesn't account for invalid input. We'll eventually need to distinguish between empty fields and invalid input.
* It would be nice to sanitize the input, e.g. to not allow entering letters in the number fields.
* Tapping the Done button only navigates back; it doesn't yet create/use the custom package.

## Changes

I used a different approach here than our validation on other screens, because we're starting with invalid values in all the fields (since they're all empty). To avoid having the validation errors appear immediately, I decided to use SwiftUI's `onReceive` to trigger the validation for each field once when the field is edited. I relied heavily on [this suggested approach](https://newcombe.io/2020/03/05/validation-with-swiftui-combine-part-1/) for inspiration and am open to suggestions for doing this differently if there are other ideas for this.

* Introduces a set of reusable publishers in `Publisher+Validation.swift` to validate if a string has content, has a double value greater than a given number, or has a double value greater than or equal to a given number.
* In `ShippingLabelCustomPackageFormViewModel`:
   * A set of validation publishers are defined, based on the reusable publishers, related to each of the fields in the form.
   * A validation publisher `packageValidation` is also defined that combines all of the validators for the individual fields. This is used to determine if the Done button should be enabled.
   * These publishers are unit tested to confirm that the validation works as expected for the input in each field in the form.
* In `ShippingLabelCustomPackageForm`:
   * Each of the text fields uses `onReceive` to trigger the appropriate validation publisher and update a property (e.g. `isNameValidated`) that tracks whether the field has valid input.
   * There is a `ValidationErrorRow` below each text field that is displayed when the input for that field is validated.
   * Does the navigation bar setup, and on any change to the entire form validates the package as a whole. This is used to determine if the Done button should be enabled.

<img src="https://user-images.githubusercontent.com/8658164/130275924-4c175c05-36ab-4a58-850a-f09362376b15.gif" width="300px">


## Testing

1. Make sure you have configured the WooCommerce WCShip plugin, and you have one or more packages set up in WP Admin under WooCommerce > Settings > Shipping > WooCommerce Shipping.
2. Launch the app in debug mode (so the feature flag is enabled).
3. Open an order detail that is eligible for the shipping label.
4. Tap the button "Create Shipping Label.”
5. Confirm "Ship From" and "Ship To.”
6. In the package details section, tap on “Package Selected.”
7. On the Package Selected screen, tap the “Create new package” button at the bottom of the screen.
8. Confirm the Add New Package screen appears and there are no validation errors yet and the Done button is disabled.
9. Start entering text in the fields, and confirm that if you remove the text you entered or enter invalid input, a validation error appears for that field.
10. Enter valid input for each field in the form, and confirm the Done button is enabled.

## Submitter Checklist

Update release notes:
